[Bonzzz3] iP

Bonzzz3 opened the pull request:
# TRACKBOT

My TrackBot project :D

> “If debugging is the process of removing software bugs, then programming must be the process of putting them in.”
― Edsger W. Dijkstra ([source](https://www.goodreads.com/quotes/1271018-if-debugging-is-the-process-of-removing-software-bugs-then)).

TrackBot frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,
1. download it from here. 
2. double-click it. 
3. add your tasks. 
4. let it manage your tasks for you 😉

And it is **FREE**!

### Features:

- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)


   ```
   Hello from TRACKBOT!
   ```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Bonzzz3 committed:
Changed main file and renamed to TrackBot

Bonzzz3 committed:
Added some initial text for chatbot

Bonzzz3 committed:
Program exits when user types "bye. Added List command too.

Bonzzz3 committed:
Updated list and add feature

Bonzzz3 committed:
Added mark as done and not done yet feature

Bonzzz3 committed:
Added todo, deadline, event and updated parts of the code

Bonzzz3 committed:
Added Text Ui Testing, expected output and input text

Bonzzz3 committed:
Added Trackbot Exceptions and updated user input formats

Bonzzz3 committed:
Updated exceptions to be more readable and added delete feature

Bonzzz3 committed:
Updated some ui

Bonzzz3 committed:
Add save feature where users can load and save data in a hard disk.

Bonzzz3 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:
  Add save feature where users can load and save data in a hard disk.

Bonzzz3 committed:
Add feature where trackbot is able to detect date and time and convert them to another format.

Bonzzz3 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add feature where trackbot is able to detect date and time and convert them to another format.

Bonzzz3 committed:
Merge branch 'master' into branch-Level-7

* master:
  Add feature where trackbot is able to detect date and time and convert them to another format.

Bonzzz3 committed:
Change expected outcome for UI test

Bonzzz3 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Change expected outcome for UI test

Bonzzz3 committed:
Make the TrackBot class more readable. Improve OOP

Bonzzz3 committed:
Merge branch 'branch-Level-7'

* branch-Level-7:

Bonzzz3 committed:
Change some file and move code to improve OOP

Bonzzz3 committed:
Move files to TrackBot package

Bonzzz3 committed:
Make the code more OOP and include packages

Bonzzz3 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Make the code more OOP and include packages

Bonzzz3 committed:
Organize classes into different packages

Bonzzz3 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Organize classes into different packages

Bonzzz3 committed:
Merge branch 'add-gradle-support'

* add-gradle-support:
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

Bonzzz3 committed:
Change gradle setting and add a simple test

Bonzzz3 committed:
Merge branch 'master' into branch-Level-8

* master:
  Change gradle setting and add a simple test
  build.gradle: Prevent generating a second JAR file
  Bump gradle and lib version
  Add Gradle support

Bonzzz3 committed:
Remove commented out code

Bonzzz3 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Remove commented out code

Bonzzz3 committed:
Add a few tests

Bonzzz3 committed:
Merge branch 'branch-Level-8'

* branch-Level-8:
  Add a few tests

Bonzzz3 committed:
Fix bye/exit command printing twice and make a jar file

Bonzzz3 committed:
Add java doc and most of the header comments

Bonzzz3 committed:
Change some access variables and comply to coding standard

Bonzzz3 committed:
Add find feature

Bonzzz3 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add java doc and most of the header comments

Bonzzz3 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Change some access variables and comply to coding standard

# Conflicts:
#	src/main/java/TrackBot/task/Deadline.java

Bonzzz3 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add find feature

# Conflicts:
#	src/main/java/TrackBot/task/TrackList.java
#	src/main/java/TrackBot/ui/Ui.java

dwsc37 commented on src/main/java/TrackBot/TrackBotStorage.java:
```diff
@@ -0,0 +1,90 @@
+ */
+public class TrackBotStorage {
+    String filePath;
+    File file;
```
I think `filePath` and `file` can have more descriptive names. For example, `taskFilePath` and `taskFile`.

dwsc37 commented on src/main/java/TrackBot/task/Deadline.java:
```diff
@@ -0,0 +1,35 @@
+ * Deadline task with deadline.
+ */
+public class Deadline extends Task {
+    private String by;
```
Again, `by` could be more descriptive, e.g. `endDate` or `dateBy`.

dwsc37 commented on src/main/java/TrackBot/task/Event.java:
```diff
@@ -0,0 +1,45 @@
+ * Event task with a start and end time.
+ */
+public class Event extends Task {
+    private String from;
+    private String to;
```
Ditto, maybe `startDate` or `dateFrom` might be better?

dwsc37 commented on src/main/java/TrackBot/task/Task.java:
```diff
@@ -0,0 +1,58 @@
+ */
+public abstract class Task {
+    protected String desc;
+    protected boolean isDone;
```
Good naming of boolean variable!

dwsc37 commented on src/main/java/TrackBot/task/TrackList.java:
```diff
@@ -0,0 +1,171 @@
+ * A list of tasks loaded from TrackBot Storage.
+ */
+public class TrackList {
+    private List<Task> list;
```
I think `tasks` might be more appropriate for a collection name?

Bonzzz3 replied:
You're right. Thanks for your review!

dwsc37 commented on src/main/java/TrackBot/ui/Parser.java:
```diff
@@ -0,0 +1,206 @@
+ */
+public class Parser {
+
+    static final String DEFAULT_DATE_FORMAT = "MMM d yyyy";
+    static final String DEFAULT_DATE_TIME_FORMAT = "MMM d yyyy HH:mm";
```
Good naming format for constants!

dwsc37 reviewed (commented):
LGTM! Naming quality is overall very good.

Bonzzz3 committed:
Add descriptions in readme file

Bonzzz3 committed:
Add checkstyle

Bonzzz3 committed:
Merge branch 'branch-Level-9'

* branch-Level-9:
  Add checkstyle

Bonzzz3 committed:
Change all files to adhere to the style guide

lutfir-cpu commented on src/main/java/trackbot/ui/Parser.java:
```diff
@@ -0,0 +1,205 @@
+import trackbot.commands.DeleteCommand;
+import trackbot.commands.ExitCommand;
+import trackbot.commands.FindCommand;
+import trackbot.commands.ListCommand;
```
All classes imported stated explicitly, no wildcard is used. Good Job!

lutfir-cpu commented on src/main/java/trackbot/commands/ExitCommand.java:
```diff
@@ -0,0 +1,25 @@
+     *
+     * @return true
+     */
+    @Override
```
Good use of naming conventions for boolean methods.

lutfir-cpu commented on src/main/java/trackbot/TrackBotStorage.java:
```diff
@@ -0,0 +1,90 @@
+     * @throws IOException If an I/O error occurs while writing to the file.
+     */
+    public void saveContents(List<Task> list) throws IOException {
+        FileWriter fw = new FileWriter(this.filePath);
```
Alternatively, naming 'fw' as 'filewriter' could make the code more readable.

lutfir-cpu commented:
LGTM! Other than a few nicks, code follows code standards very closely.

lutfir-cpu commented on src/main/java/trackbot/ui/Parser.java:
```diff
@@ -0,0 +1,205 @@
+        String commandWord = splitInput[0].toLowerCase();
+        String otherWord = splitInput.length > 1 ? splitInput[1].trim() : "";
+
+        switch (commandWord) {
```
Switch statements shouldn't have extra indentations. Personally I would have done something like this:

``` 
switch (commandWord) {
case "list": 
if (!otherWord.isEmpty()) {
    throw TrackBotException.invalidFormat("list", "list");
}
return new ListCommand();
.
.
.
```

Bonzzz3 replied:
I believe what I did was correct . The below code is taken from the java coding standard given [here](https://se-education.org/guides/conventions/java/intermediate.html)


```ruby
switch (condition) {
case ABC:
    statements;
    // Fallthrough
case DEF:
    statements;
    break;
case XYZ:
    statements;
    break;
default:
    statements;
    break;
}
```

lutfir-cpu replied:
You're right, misread the coding standard. Thanks.

lutfir-cpu commented on src/main/java/trackbot/ui/Parser.java:
```diff
@@ -0,0 +1,205 @@
+        case "T":
+            task = new ToDo(parts[2]);
+            break;
+        case "D":
```
Switch code block shouldn't have an extra indentation, refer to previous comment for example.

Bonzzz3 committed:
Add simple javafx to have GUI

Bonzzz3 committed:
Add all valid response GUI according to user input

Runus01 commented on src/main/java/trackbot/TrackBotStorage.java:
```diff
@@ -0,0 +1,90 @@
+import java.io.FileNotFoundException;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.util.ArrayList;
```
I put a blank line to split up the java.io and java.util packages but that's just a personal preference.

Runus01 commented on src/main/java/trackbot/task/Event.java:
```diff
@@ -0,0 +1,44 @@
+ * Event task with a start and end time.
+ */
+public class Event extends Task {
+    private String from;
```
Perhaps a more descriptive name for these fields might be nice such as `startDate` and `endDate`.

Runus01 commented on src/main/java/trackbot/task/ToDo.java:
```diff
@@ -0,0 +1,26 @@
+    }
+
+    @Override
+    public String toStorageFormat() {
```
Great idea to have each class handle the conversion of the Task object to a String for storage in the data file. I had my main class handle this conversion which made things a bit complex.

Runus01 commented on src/main/java/trackbot/task/TrackList.java:
```diff
@@ -0,0 +1,171 @@
+    private final TrackBotStorage storage;
+
+    /**
+     * Constructs a TrackList and loads tasks from the specified storage.
```
Great use of javadoc comments to explain what every method and class does.

Runus01 commented on src/test/java/trackbot/ui/ParserTest.java:
```diff
@@ -0,0 +1,86 @@
+public class ParserTest {
+    private final Parser parser = new Parser();
+
+    @Test
```
Good use of a consistent format to write tests. They are very readable and easy to follow.

Bonzzz3 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Add all valid response GUI according to user input
  Add simple javafx to have GUI

Bonzzz3 committed:
Change exceptions to visible response to user

No visible response is shown to reply to user's invalid input.

Adding a new class, InvalidCommand which can execute invalid format commands allow response to be shown.

Also moved some files into a single gui folder to avoid messy code.

Bonzzz3 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Change exceptions to visible response to user

Bonzzz3 committed:
Merge branch 'master' into branch-A-CodingStandard

* master:
  Change exceptions to visible response to user
  Add all valid response GUI according to user input
  Add simple javafx to have GUI
  Change all files to adhere to the style guide
  Add checkstyle
  Add descriptions in readme file
  Add find feature
  Add java doc and most of the header comments

Bonzzz3 committed:
Update code based on peer review suggestions

Bonzzz3 committed:
Merge branch 'branch-A-CodingStandard'

* branch-A-CodingStandard:
  Update code based on peer review suggestions

Bonzzz3 committed:
Merge branch 'master' into branch-Level-10

* master:
  Update code based on peer review suggestions

Bonzzz3 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:

Bonzzz3 committed:
Add assertions
To verify assumptions about internal invariants.

Bonzzz3 committed:
Add more assertion
With more assertion added and message to show what go wrong, it is easier for us to spot programmer's mistake.

Bonzzz3 committed:
Merge pull request #1 from Bonzzz3/branch-A-Assertions

Branch a assertions

Bonzzz3 committed:
Merge branch 'master' of https://github.com/Bonzzz3/ip

* 'master' of https://github.com/Bonzzz3/ip:
  Add more assertion With more assertion added and message to show what go wrong, it is easier for us to spot programmer's mistake.
  Add assertions To verify assumptions about internal invariants.

Bonzzz3 committed:
Merge branch 'master' into branch-A-Streams

* master:
  Add more assertion With more assertion added and message to show what go wrong, it is easier for us to spot programmer's mistake.
  Add assertions To verify assumptions about internal invariants.

Bonzzz3 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add more assertion With more assertion added and message to show what go wrong, it is easier for us to spot programmer's mistake.
  Add assertions To verify assumptions about internal invariants.

Bonzzz3 committed:
Improve code quality

Improve readability of the code so that it can be easily understood and modified in the future.

Bonzzz3 committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Improve code quality

Bonzzz3 committed:
Merge branch 'branch-A-Streams'

* branch-A-Streams:

Bonzzz3 committed:
Add C-DetectDuplicates

Improves user experience by ensuring that tasks with the same description are not added multiple times.

Bonzzz3 committed:
Merge branch 'branch-A-Extensions'

* branch-A-Extensions:
  Add C-DetectDuplicates

Bonzzz3 committed:
Merge branch 'master' into branch-A-CodeQuality

* master:
  Add C-DetectDuplicates

Bonzzz3 committed:
Improve code quality

Improves readability and cleans up code.

Deleting commented out code and correcting the format for java comment
header makes code clean.

Bonzzz3 committed:
Improve readability

Removing deep nesting codes and minimising code duplication helps
reader to keep track of the logic and improve quality of the code.

Let's add more constants and extract duplicated codes as a method
and rename variables.

Bonzzz3 committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Improve readability
  Improve code quality

Bonzzz3 committed:
Add Javadoc comments

Describes what each method do and their return value which improves
readability.

Let's add missing Javadoc comments to some methods in the ui class.

Bonzzz3 committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Add Javadoc comments

Bonzzz3 committed:
Add title

Adding product name on the title bar so that it is visible to users.

Bonzzz3 committed:
Merge branch 'branch-A-CodeQuality'

* branch-A-CodeQuality:
  Add title

Bonzzz3 committed:
Merge branch 'master' into branch-Level-10

* master:
  Add title
  Add Javadoc comments
  Improve readability
  Improve code quality
  Add C-DetectDuplicates
  Improve code quality
  Add more assertion With more assertion added and message to show what go wrong, it is easier for us to spot programmer's mistake.
  Add assertions To verify assumptions about internal invariants.

Bonzzz3 committed:
Update GUI and fix minor bug

Setting text field, button and scroll pane to automatically resize
improves user experience when using the app.

Let's change the minor mistake of the regex year format.

Bonzzz3 committed:
Merge branch 'branch-Level-10'

* branch-Level-10:
  Update GUI and fix minor bug

Bonzzz3 committed:
Add ui

Adding a screenshot of ui to show how the app looks like.

Bonzzz3 committed:
Update README for user guide

Updating the product name and adding a some features descriptions
so that users can follow it to use the app.

Bonzzz3 committed:
Merge branch 'master' into branch-A-JavaDoc

* master:
  Update README for user guide
  Add ui
  Update GUI and fix minor bug
  Add title
  Add Javadoc comments
  Improve readability
  Improve code quality
  Add C-DetectDuplicates
  Improve code quality
  Add more assertion With more assertion added and message to show what go wrong, it is easier for us to spot programmer's mistake.
  Add assertions To verify assumptions about internal invariants.
  Update code based on peer review suggestions
  Change exceptions to visible response to user
  Add all valid response GUI according to user input
  Add simple javafx to have GUI
  Change all files to adhere to the style guide
  Add checkstyle
  Add descriptions in readme file
  Add find feature
  Change some access variables and comply to coding standard

Bonzzz3 committed:
Add more features descriptions to README

Adding an updated Ui of the app and updated features for user guide.

Bonzzz3 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Add more features descriptions to README

Bonzzz3 committed:
Update README for user guide

Adding command summary for users to have a quick view of commands.

Bonzzz3 committed:
Merge branch 'branch-A-JavaDoc'

* branch-A-JavaDoc:
  Update README for user guide

Bonzzz3 committed:
Fix bug

Changing how system exit so that it can exit on all platforms.

Bonzzz3 committed:
Update README for user guide

Remove redundant code.
